qiskit_experiments/calibration_management/base_calibration_experiment.py

```diff
@@ -0,0 +1,90 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Base class for calibration-type experiments."""
+
+from abc import abstractmethod
+from typing import Optional
+
+from qiskit.providers.options import Options
+from qiskit.providers.backend import Backend
+
+from qiskit_experiments.framework.base_experiment import BaseExperiment
+from qiskit_experiments.framework.experiment_data import ExperimentData
+
+
+class BaseCalibrationExperiment(BaseExperiment):
+    """An abstract base class for calibration experiments.
+
+    This abstract base class specifies an experiment and how to update an
+    optional instance of :class:`Calibrations` specified in the experiment options
+    under calibrations. Furthermore, the experiment options also specifies
+    an auto_update variable which, by default, is set to True. If this variable,
+    is True then the run method of the experiment will call :meth:`block_for_results`
+    and update the calibrations instance.
+    """
+
+    # The updater class that updates the Calibrations instance
+    __updater__ = None
+
+    @abstractmethod
+    def update_calibrations(self, experiment_data: ExperimentData):
+        """Update parameter values in the :class:`Calibrations` instance.
+
+        Subclasses must implement this method which will call the :meth:`update`
+        method of the updater.
+        """
+
+    @classmethod
+    def _default_experiment_options(cls) -> Options:
+        """Default options for experiment
+
+        Experiment Options:
+            calibrations (Calibrations): An optional instance of :class:`Calibrations` if this
+                instance is specified then the experiment will try and update the calibrations.
+            auto_update (bool): A boolean which defaults to True. If this variable is set to
+                True then running the calibration experiment will block for the results and
+                update the calibrations if the calibrations is not None.
+        """
+        options = super()._default_experiment_options()
+        options.calibrations = None
+        options.auto_update = True
+
+        return options
+
+    def run(
```

Comment:
I'm not sure what you mean by

> would it be better to provide an extra hook in the base class that run() calls?

can you clarify?

Reply:
I think this can be done in the separate PR. Currently we have `post_transpile` method as an extra hook. It's good time to think about a mechanism to manage these hooks.

Reply:
I meant that `BaseExperiment` currently provides several methods that subclasses can implement like `circuits()`, `analysis()`, and `_run_analysis()` (in `BaseAnalysis`) to modify behavior without needing to override the base `run()` method. Could there be another method that `BaseExperiment.run()` could call after the analysis to do the update? The futures-like aspect of analysis results makes this difficult -- it would need to be a callback on the analysis callback.

Or the calibration update could be part of the analysis. The downside to that to me is that then it would try to update the `Calibrations` if one re-ran the analysis which might be unexpected behavior.

Reply:
Initially I wanted to do the update in the analysis class. However, the update requires the metadata in the `experiment_data` instance. Therefore we cannot use `analysis_result` alone as it does not contain all the needed information. Furthermore, I did not find any obvious place in `BaseAnalysis` or `CurveAnalysis` to have something like
```
def update_calibrations(self, experiment_data: ExperimentData)
```
which also needs to know the calibrations. (Perhaps you see something? I did not.). 

Now, looking at `BaseExperiment` it looks like we could override `run_analysis` instead of `run`. I think this would be less disruptive than overriding `run`. How does that sound to you?

Reply:
I still prefer separating analysis and calibration because we don't want to assume calibration in base experiment. I rather prefer analysis taking callback (but this also means calibration is a part of analysis), but I think the best solution is to introduce hook mechanism. For example, qiskit machine learning has pre/post hook to each processing step.

Reply:
In `BaseAnalysis` right before the return statement of `run` we could do `self._post_analysis(experiment_data, **options)` and use that to do the updating. But I'd rather do the updating in a hook in the subclasses of `BaseCalibrationExperiment` as it keeps all the calibration stuff in one place. If we do it in the analysis class we also need to pass the calibrations to the analysis classes. This will generate more overhead and make it more complex to implement future experiments.